### PR TITLE
Known hosts locking

### LINKED ISSUE #2814: known_hosts file prone to synchronisation errors
When using multiple FreeRDP instances the ```known_hosts``` file is prone to synchronous modification issues.

Need to implement file locking to ensure the file is not modified in an illegal way (ending up with duplicate entries)

### PR DESCRIPTION
**Pull request by akallabeth**
Fixes #2814:
* Implements known_hosts file operations using windows style API with locking.
* Implements windows style API functions for files in winpr
* Implements (rudimentary) file locking with ```CreateFile```

### CODE REVIEW

File: winpr/libwinpr/utils/debug.c [hunk 1]
```diff
@@ -471,3 +471,29 @@ void winpr_log_backtrace(const char* tag, DWORD level, DWORD size)
 	winpr_backtrace_free(stack);
 }
 
+char* winpr_strerror(DWORD dw, char* dmsg, size_t size) {
```
Comment (bmiklautz) [inline winpr/libwinpr/utils/debug.c]: Where is this function used?

**Comment by bmiklautz** [comment]
@akallabeth is it possible you add some basic tests for at least the file locking?

File: winpr/libwinpr/utils/debug.c (hunk 1 above)
Comment (akallabeth) [inline winpr/libwinpr/utils/debug.c]: Currently unused, had implemented it while debugging.
But is quite handy so I left it in there, can easily be used as platform independent version of ```strerror```

**Comment by akallabeth** [comment]
@bmiklautz @awakecoding @hardening Rebased the whole pr to current master. Would greatly imporve the file functions of winpr, could you have a look?

File: libfreerdp/crypto/certificate.c [hunk 2]
```diff
@@ -415,22 +395,26 @@ BOOL certificate_data_replace(rdpCertificateStore* certificate_store,
...
 	}
 
-	fclose(fp);
-
-	fp = fopen(certificate_store->file, "wb");
+	if (SetFilePointer(fp, 0, NULL, FILE_BEGIN) == INVALID_SET_FILE_POINTER)
+	{
+		WLog_ERR(TAG, "SetFilePointer(%s) returned %s [%08X]",
+			 certificate_store->file, strerror(errno), GetLastError());
+		CloseHandle(fp);
```
Comment (mfleisz) [inline libfreerdp/crypto/certificate.c]: leaking "data" here and in all subsequent return FALSE branches

**Comment by mfleisz** [comment]
Despite my remark +1 on this

File: winpr/libwinpr/file/file.c [hunk 3]
```diff
@@ -159,29 +223,349 @@ static BOOL FileWrite(PVOID Object, LPCVOID lpBuffer, DWORD nNumberOfBytesToWrit
...
+		if (!fp)
+		{
+			free(pFile->lpFileName);
+			free(pFile);
+			return INVALID_HANDLE_VALUE;
+		}
+		fclose(fp);
+	}
+
+	{
```
Comment (hardening) [inline winpr/libwinpr/file/file.c]: isn't possible that we have a race condition here. If another process is fast enough to delete pFile->lpFileName before it is reopened ?

**Comment by hardening** [comment]
Except my remark +1.
This is really a nice, big work, congratulations !

**Comment by bmiklautz** [comment]
Although the pull request title doesn't give it really away this is rather a big change (and improvement) in file handling. 
@akallabeth really nicely done! 

+1

**Pull request merged**